File: backend/app/routers/venues.py
import datetime
from zoneinfo import ZoneInfo
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import or_, and_

from ..database import get_db
from ..models import Venue, VenueBusynessHourly
from ..schemas import VenueSummary, VenueDetail, VenueListResponse

router = APIRouter(prefix="/venues", tags=["venues"])

SORT_FIELDS = {
    "rankingScore": Venue.ranking_score,
    "quietScore": Venue.quiet_score,
    "mentionCount": Venue.mention_count,
}


def _crowdedness_from_busyness(pct: float | None) -> str | None:
    """Derive crowdedness from ML busyness percentage.
    Calibrated against observed data: peak avg ~18, peak max ~64.
    """
    if pct is None:
        return None
    if pct < 8:
        return "low"
    if pct < 25:
        return "medium"
    return "high"


def _crowdedness_from_scores(attribute_scores: dict | None) -> str | None:
    """Fallback: derive crowdedness from editorial attribute scores."""
    if not attribute_scores:
        return None
    crowding = attribute_scores.get("crowding", {})
    if not crowding or crowding.get("n", 0) < 2:
        return None
    net = crowding.get("net", 0)
    if net >= 1:
        return "low"
    if net <= -2:
        return "high"
    return "medium"


def _osm_tag(v: Venue, key: str) -> str | None:
    """Raw OSM tag value, or None when the venue has no OSM match."""
    if v.osm and isinstance(v.osm.get("tags"), dict):
        return v.osm["tags"].get(key)
    return None


def _osm_opening_hours(v: Venue) -> str | None:
    return _osm_tag(v, "opening_hours")


def _wheelchair_access(v: Venue) -> str | None:
    """Wheelchair access. Hand-verified `accessibility_manual` beats the OSM tag."""
    manual = v.accessibility_manual or {}
    if isinstance(manual, dict) and manual.get("wheelchair") in {"yes", "limited", "no"}:
        return manual["wheelchair"]
    tag = _osm_tag(v, "wheelchair")
    return tag if tag in {"yes", "limited", "no"} else None


def venue_to_summary(v: Venue, busyness_pct: float | None = None) -> VenueSummary:
    # Prefer ML busyness data; fall back to editorial crowding scores
    if busyness_pct is not None:
        crowdedness = _crowdedness_from_busyness(busyness_pct)
    else:
        crowdedness = _crowdedness_from_scores(v.attribute_scores)

    return VenueSummary(
        id=v.canonical_id,
        placeId=v.place_id,
        name=v.canonical_name,
        address=v.formatted_address or "",
        type=v.venue_type or "other",
        lat=v.lat or 0.0,
        lng=v.lng or 0.0,
        quietScore=v.quiet_score or 0,
        displayRating=v.display_rating or 0.0,
        rankingScore=v.ranking_score or 0.0,
        mentionCount=v.mention_count or 0,
        uniqueSources=v.unique_sources or 0,
        businessStatus=v.business_status or "OPERATIONAL",
        crowdedness=crowdedness,
        hours=_osm_opening_hours(v),
        wheelchair=_wheelchair_access(v),
    )


def venue_to_detail(v: Venue, busyness_pct: float | None = None) -> VenueDetail:
    # `hours` now comes through the summary, so it must not be passed again here.
    return VenueDetail(
        **venue_to_summary(v, busyness_pct).model_dump(),
        attributeScores=v.attribute_scores or {},
        representativeQuotes=v.representative_quotes or [],
        osm=v.osm,
        accessibilityManual=v.accessibility_manual,
    )


_NYC_TZ = ZoneInfo("America/New_York")


def _fetch_busyness_map(db: Session, venue_ids: list[str]) -> dict[str, float]:
    """Batch-fetch current-hour busyness_pct for a list of venue IDs.
    Uses NYC local time since venue busyness data is keyed to NYC timezone.
    """
    now = datetime.datetime.now(_NYC_TZ)
    rows = (
        db.query(VenueBusynessHourly)
        .filter(
            VenueBusynessHourly.canonical_id.in_(venue_ids),
            VenueBusynessHourly.dow == now.weekday(),
            VenueBusynessHourly.hour == now.hour,
        )
        .all()
    )
    return {r.canonical_id: r.busyness_pct for r in rows if r.busyness_pct is not None}


@router.get("", response_model=VenueListResponse)
def list_venues(
    q: Optional[str] = Query(None, description="Search name or address"),
    type: Optional[str] = Query(None, description="Comma-separated: cafe,library,restaurant,bar"),
    minQuiet: Optional[int] = Query(None, ge=0, le=100),
    bounds: Optional[str] = Query(None, description="swLat,swLng,neLat,neLng"),
    sort: str = Query("rankingScore", description="rankingScore|quietScore|mentionCount"),
    minMentions: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    query = db.query(Venue)

    # Text search
    if q:
        term = f"%{q}%"
        query = query.filter(
            or_(
                Venue.canonical_name.ilike(term),
                Venue.formatted_address.ilike(term),
            )
        )

    # Type filter
    if type:
        types = [t.strip() for t in type.split(",")]
        query = query.filter(Venue.venue_type.in_(types))

    # Quiet score filter
    if minQuiet is not None:
        query = query.filter(Venue.quiet_score >= minQuiet)

    # Mention count filter
    query = query.filter(Venue.mention_count >= minMentions)

    # Map bounds filter
    if bounds:
        try:
            sw_lat, sw_lng, ne_lat, ne_lng = map(float, bounds.split(","))
            query = query.filter(
                and_(
                    Venue.lat >= sw_lat,
                    Venue.lat <= ne_lat,
                    Venue.lng >= sw_lng,
                    Venue.lng <= ne_lng,
                )
            )
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid bounds format. Use: swLat,swLng,neLat,neLng")

    total = query.count()

    # Sorting
    sort_col = SORT_FIELDS.get(sort, Venue.ranking_score)
    query = query.order_by(sort_col.desc())

    venues = query.offset(offset).limit(limit).all()

    # Batch-fetch current-hour busyness for all venues in one query
    venue_ids = [v.canonical_id for v in venues]
    busyness_map = _fetch_busyness_map(db, venue_ids)

    return VenueListResponse(
        items=[venue_to_summary(v, busyness_map.get(v.canonical_id)) for v in venues],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/{venue_id}", response_model=VenueDetail)
def get_venue(venue_id: str, db: Session = Depends(get_db)):
    venue = db.query(Venue).filter(Venue.canonical_id == venue_id).first()
    if not venue:
        raise HTTPException(status_code=404, detail="Venue not found")

    # Fetch current-hour busyness for this single venue
    busyness_map = _fetch_busyness_map(db, [venue_id])
    return venue_to_detail(venue, busyness_map.get(venue_id))

File: frontend/src/services/venues.js
import { apiRequest } from './api'
import { fetchVenueSignals, resolveVenueSignals } from './venueSignals'
import {
  busynessPctToLevel,
  getBusynessCrowdednessLevel,
  quietInsideLevelFromSignals,
} from '../utils/venueSignalsDisplay'
import { getBusyLevelNowFromBusyness } from '../utils/busynessForecast'
import { applyUserRatingFields, mergeGoogleRatingIntoPlace } from '../utils/userRating'
import { fetchPlaceCoreDetails, getCachedPlaceCoreDetails } from './googleMapsData'

const DEFAULT_MIN_MENTIONS = 1

const PHOTO_SETS = {
  cafe: [
    'https://images.unsplash.com/photo-1501339847302-ac426a4a7cbb?w=800&q=80',
  ],
  library: [
    'https://images.unsplash.com/photo-1481627834876-b7833e8f5570?w=800&q=80',
    'https://images.unsplash.com/photo-1521587760476-6c12a4b040da?w=800&q=80',
  ],
  public: [
    'https://images.unsplash.com/photo-1441974231531-c6227db76b6e?w=800&q=80',
  ],
}

export function mapBackendType(type, name = '') {
  if (type === 'library') return 'library'
  // Dataset often labels cafés as restaurant/bar.
  if (type === 'cafe' || type === 'restaurant' || type === 'bar') return 'cafe'
  if (type === 'public' || type === 'park') return 'public'

  // Ambiguous labels (other / hotel / …) — infer from name.
  const n = String(name || '').toLowerCase()
  if (/\b(librar|reading room)\b/.test(n)) return 'library'
  if (
    /\b(cafe|café|coffee|bakery|bakeshop|espresso|tea house|tea shop)\b/.test(n)
  ) {
    return 'cafe'
  }

  return 'public'
}

function mapSummaryCrowdedness(summary) {
  // Prefer list-payload crowdedness (already batch-derived from current-hour
  // busyness on the backend). quiet-profile enrichment can refine later.
  const raw = summary.busynessLevel ?? summary.busyness_level ?? summary.crowdedness
  if (raw === 'moderate') return 'medium'
  if (['low', 'medium', 'high'].includes(raw)) return raw
  return null
}

function crowdingFromAttributes(attributeScores) {
  const crowding = attributeScores?.crowding
  if (!crowding || crowding.n < 2) return null
  if (crowding.net >= 1) return 'low'
  if (crowding.net <= -2) return 'high'
  return 'medium'
}

function buildTags({ mentionCount, attributeScores, osm, accessibilityManual }) {
  const tags = new Set()

  if (mentionCount > 0) {
    tags.add(`${mentionCount} write-ups`)
  }

  for (const [key, value] of Object.entries(attributeScores ?? {})) {
    if (!value || value.n < 2 || !value.note || value.net <= 0) continue
    tags.add(`${key}: ${value.note}`)
  }

  const osmTags = osm?.tags
  if (osmTags?.internet_access) tags.add('WiFi available')
  if (osmTags?.wheelchair === 'yes' || accessibilityManual?.wheelchair === 'yes') {
    tags.add('Wheelchair accessible')
  }
  if (osmTags?.outdoor_seating) tags.add('Outdoor seating')

  return Array.from(tags).slice(0, 6)
}

function latLngToMapPercent(lat, lng) {
  const minLat = 40.704
  const maxLat = 40.882
  const minLng = -74.02
  const maxLng = -73.907

  const mapX = ((lng - minLng) / (maxLng - minLng)) * 80 + 10
  const mapY = ((maxLat - lat) / (maxLat - minLat)) * 70 + 10

  return {
    mapX: Math.max(5, Math.min(95, mapX)),
    mapY: Math.max(5, Math.min(95, mapY)),
  }
}

export function apiSummaryToPlace(summary) {
  const type = mapBackendType(summary.type, summary.name)
  const quietScore = summary.quietScore ?? 0
  const displayRating = summary.displayRating ?? 0
  const userRatingFields = applyUserRatingFields(summary)
  const { mapX, mapY } = latLngToMapPercent(summary.lat, summary.lng)
  const zipMatch = (summary.address ?? '').match(/\b(10\d{3})\b/)

  return {
    id: summary.id,
    placeId: summary.placeId,
    name: summary.name,
    canonicalName: summary.name,
    address: summary.address,
    zipCode: zipMatch?.[1] ?? '10001',
    type,
    lat: summary.lat,
    lng: summary.lng,
    mapX,
    mapY,
    quietScore,
    rankingScore: summary.rankingScore ?? 0,
    mentionCount: summary.mentionCount ?? 0,
    uniqueSources: summary.uniqueSources ?? 0,
    displayRating,
    ...userRatingFields,
    occupancy: null,
    capacity: 0,
    currentPct: summary.currentPct ?? summary.current_pct ?? null,
    // List API already batch-computes this from current-hour busyness.
    crowdedness: mapSummaryCrowdedness(summary),
    // 'calm' | 'mixed' | 'lively' | null (null = no reviewer mentioned it).
    // Filled lazily via enrichPlacesFromQuietProfile when the filter is used.
    quietInside: null,
    // Fallback source: raw OSM `opening_hours` string from our own venue data.
    // Only present once the backend exposes `hours` on VenueSummary.
    openingHours: summary.hours ?? null,
    // Primary source: Google Places, filled by enrichPlacesFromGoogleRatings.
    openingPeriods: null,
    openingWeekdayText: null,
    // OSM `wheelchair` tag: 'yes' | 'limited' | 'no' | null. Requires the
    // backend to expose `wheelchair` on VenueSummary; detail payloads carry it
    // in the raw osm blob regardless.
    wheelchairAccess: summary.wheelchair ?? null,
    googleRating: null,
    googleRatingCount: 0,
    distance: 0,
    hours: 'Hours unavailable',
    tags: (summary.mentionCount ?? 0) > 0 ? [`${summary.mentionCount} write-ups`] : [],
    photos: PHOTO_SETS[type] ?? PHOTO_SETS.public,
    editorialQuotes: [],
    businessStatus: summary.businessStatus ?? 'OPERATIONAL',
    attributeScores: {},
    osm: null,
    accessibilityManual: null,
  }
}

function enrichPlaceWithDetail(place, detail, signals, live) {
  const type = mapBackendType(detail.type, detail.name ?? place?.name)
  const attributeScores = detail.attributeScores ?? {}
  const tags = buildTags({
    mentionCount: detail.mentionCount,
    attributeScores,
    osm: detail.osm,
    accessibilityManual: detail.accessibilityManual,
  })

  const resolvedSignals = resolveVenueSignals(signals, { ...place, attributeScores })
  // Match Busy level chart "now" (hourly curve), not raw current_pct alone.
  const busyNow = getBusyLevelNowFromBusyness(resolvedSignals?.busyness)
  const busynessLevel =
    busyNow?.level ?? getBusynessCrowdednessLevel(resolvedSignals?.busyness)
  const mapped = apiSummaryToPlace(detail)

  // Detail payloads often omit user ratings; don't wipe list/review values on click.
  const detailHasUserRating =
    detail.userRating != null && (detail.userReviewCount ?? 0) > 0

  return {
    ...place,
    ...mapped,
    type,
    hours: detail.hours ?? place.hours,
    // Detail always carries the raw OSM string, so a venue opened from the list
    // gets real hours even before the list payload exposes them.
    openingHours: detail.hours ?? place.openingHours ?? null,
    // `mapped` is built from the detail payload and resets these to null, so
    // the already-enriched Google values have to be restored explicitly.
    openingPeriods: place.openingPeriods ?? null,
    openingWeekdayText: place.openingWeekdayText ?? null,
    wheelchairAccess:
      detail.osm?.tags?.wheelchair ?? place.wheelchairAccess ?? null,
    tags: tags.length ? tags : place.tags,
    editorialQuotes: (detail.representativeQuotes ?? []).map((quote, index) => ({
      publication: quote.publication,
      text: quote.quote,
      sourceUrl: quote.source_url,
      order: index + 1,
    })),
    attributeScores,
    osm: detail.osm,
    accessibilityManual: detail.accessibilityManual,
    capacity: Number(detail.osm?.tags?.capacity) || place.capacity,
    currentPct:
      busyNow?.busynessPct ??
      resolvedSignals?.busyness?.currentBusynessPct ??
      mapped.currentPct ??
      place.currentPct ??
      null,
    // Same level as right-panel Busy level "now".
    crowdedness:
      busynessLevel ??
      mapSummaryCrowdedness(detail) ??
      live?.crowdedness ??
      crowdingFromAttributes(attributeScores) ??
      place.crowdedness,
    occupancy: live?.occupancy ?? null,
    quietInside:
      quietInsideLevelFromSignals(resolvedSignals) ?? place.quietInside ?? null,
    signals: resolvedSignals,
    userRating: detailHasUserRating ? mapped.userRating : (place.userRating ?? mapped.userRating),
    userReviewCount: detailHasUserRating
      ? mapped.userReviewCount
      : (place.userReviewCount ?? mapped.userReviewCount),
    rating: detailHasUserRating ? mapped.rating : (place.rating ?? mapped.rating),
    // Google stars come from Places API enrich, not venue detail payload.
    googleRating: place.googleRating ?? mapped.googleRating,
    googleRatingCount: place.googleRatingCount ?? mapped.googleRatingCount ?? 0,
  }
}

export async function fetchVenueDetail(venueId, basePlace, language) {
  const langQuery = language ? `?lang=${encodeURIComponent(language)}` : ''
  const [detail, signals, live] = await Promise.all([
    apiRequest(`/venues/${encodeURIComponent(venueId)}${langQuery}`),
    fetchVenueSignals(venueId),
    apiRequest(`/venues/${encodeURIComponent(venueId)}/live`).catch(() => null),
  ])

  return enrichPlaceWithDetail(basePlace, detail, signals, live)
}

/**
 * Fill list-card crowdedness / indoor-quiet from quiet-profile.
 * Prefer passing a subset via `placeIds` — a full 326-venue fan-out is slow.
 * Calls onPlaceUpdate with a new place object (immutable) as each venue resolves.
 */
export async function enrichPlacesFromQuietProfile(
  places,
  { concurrency = 4, onPlaceUpdate, placeIds = null } = {},
) {
  if (!places?.length) return places

  const allow = placeIds ? new Set(placeIds) : null
  const queue = places
    .map((place, index) => ({ place, index }))
    .filter(({ place }) => !allow || allow.has(place.id))

  if (!queue.length) return places

  let cursor = 0
  const workers = Array.from({ length: Math.min(concurrency, queue.length) }, async () => {
    while (cursor < queue.length) {
      const i = cursor
      cursor += 1
      const { place, index } = queue[i]
      try {
        const signals = await fetchVenueSignals(place.id)
        const busyNow = getBusyLevelNowFromBusyness(signals?.busyness)

        let currentPct = place.currentPct ?? null
        let crowdedness = place.crowdedness ?? null

        if (busyNow?.busynessPct != null) currentPct = busyNow.busynessPct
        if (busyNow?.level) {
          crowdedness = busyNow.level
        } else {
          const pct = signals?.busyness?.currentBusynessPct
          const level = busynessPctToLevel(pct) ?? getBusynessCrowdednessLevel(signals?.busyness)
          if (pct != null) currentPct = pct
          if (level) crowdedness = level
        }

        // Same request already carries calm_bar, so the indoor-quiet level for
        // the list filter costs no extra round trip.
        const quietInside =
          quietInsideLevelFromSignals(signals) ?? place.quietInside ?? null

        if (
          crowdedness === place.crowdedness &&
          currentPct === place.currentPct &&
          quietInside === place.quietInside
        ) {
          continue
        }

        const updated = { ...place, currentPct, crowdedness, quietInside }
        places[index] = updated
        onPlaceUpdate?.(updated)
      } catch {
        // Keep place without badge if quiet-profile fails.
      }
    }
  })

  await Promise.all(workers)
  return places
}

/**
 * Apply cached Google rating/hours onto places synchronously so the list can
 * paint stars on the first frame after /venues returns.
 */
export function hydratePlacesFromGoogleCache(places) {
  if (!places?.length) return places

  for (let i = 0; i < places.length; i += 1) {
    const place = places[i]
    if (!place?.placeId) continue
    if (place.googleRating != null && place.openingPeriods != null) continue

    const cached = getCachedPlaceCoreDetails(place.placeId)
    if (!cached) continue

    const withRating =
      cached.rating != null
        ? mergeGoogleRatingIntoPlace(place, {
            rating: cached.rating,
            userRatingCount: cached.userRatingCount,
          })
        : place

    places[i] = {
      ...withRating,
      openingPeriods: cached.openingPeriods ?? place.openingPeriods,
      openingWeekdayText: cached.openingWeekdayText ?? place.openingWeekdayText,
    }
  }

  return places
}

/**
 * Fill list/detail card stars AND opening hours from Google Places (placeId).
 * Both come from a single Place Details call per venue at no extra SKU cost.
 *
 * Pass `placeIds` to enrich a priority subset first; call again without it for
 * the remainder. Default concurrency is low so the UI stays responsive.
 */
export async function enrichPlacesFromGoogleRatings(
  places,
  { concurrency = 6, onPlaceUpdate, placeIds = null } = {},
) {
  if (!places?.length) return places

  // Places library loads with the map script — wait briefly if needed.
  const readyDeadline = Date.now() + 8000
  while (Date.now() < readyDeadline) {
    if (typeof google !== 'undefined' && google.maps?.importLibrary) break
    await new Promise((resolve) => setTimeout(resolve, 100))
  }

  const allow = placeIds ? new Set(placeIds) : null
  const queue = places
    .map((place, index) => ({ place, index }))
    .filter(({ place }) => {
      if (allow && !allow.has(place.id)) return false
      if (!place.placeId) return false
      // Hours are needed even when the rating is already cached, so the guard
      // checks both rather than short-circuiting on rating alone.
      return place.googleRating == null || place.openingPeriods == null
    })

  if (!queue.length) return places

  let cursor = 0
  const workers = Array.from({ length: Math.min(concurrency, queue.length) }, async () => {
    while (cursor < queue.length) {
      const i = cursor
      cursor += 1
      const { place, index } = queue[i]

      try {
        const details = await fetchPlaceCoreDetails(place.placeId)
        if (!details) continue

        const withRating =
          details.rating != null
            ? mergeGoogleRatingIntoPlace(place, {
                rating: details.rating,
                userRatingCount: details.userRatingCount,
              })
            : place

        const updated = {
          ...withRating,
          openingPeriods: details.openingPeriods ?? place.openingPeriods,
          openingWeekdayText:
            details.openingWeekdayText ?? place.openingWeekdayText,
        }

        places[index] = updated
        onPlaceUpdate?.(updated)
      } catch {
        // Keep empty Google rating / hours if the Places request fails.
      }
    }
  })

  await Promise.all(workers)
  return places
}

export async function loadVenues(options = {}) {
  const minMentions = options.minMentions ?? DEFAULT_MIN_MENTIONS
  const language = options.language
  const limit = 200
  let offset = 0
  const all = []

  while (true) {
    const params = new URLSearchParams({
      minMentions: String(minMentions),
      limit: String(limit),
      offset: String(offset),
      sort: 'rankingScore',
    })
    if (language) {
      params.set('lang', language)
    }
    const page = await apiRequest(`/venues?${params.toString()}`)
    all.push(...(page.items ?? []))

    if (all.length >= page.total || (page.items ?? []).length < limit) {
      break
    }
    offset += limit
  }

  const places = all.map(apiSummaryToPlace)
  return places
}
